llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp

```diff
@@ -438,6 +438,73 @@ bool duplicateAdressingRegs(MachineBasicBlock &MBB, MachineRegisterInfo &MRI) {
   return tryToDuplicateLoadUse(LoadUses, NonLoadUses, MRI, TII);
 }
 
+/// This optimization tries to propagate GlobalValues as MMO of load operations
+/// that don't use normal pointers as a source of an address.
+void findGlobalValues(const MachineInstr &MI,
+                      SmallPtrSet<const Value *, 4> &GVSet,
+                      SmallPtrSet<const MachineInstr *, 4> &VisitedInstrs,
+                      unsigned RecMax, MachineRegisterInfo &MRI) {
+
+  // Exhausted the search.
+  if (RecMax == 0)
+    return;
+
+  // Loaded values cannot be tracked as Globals.
+  if (MI.mayLoad())
+    return;
+
+  if (VisitedInstrs.find(&MI) != VisitedInstrs.end())
+    return;
+
+  VisitedInstrs.insert(&MI);
+
+  for (auto MO : MI.uses()) {
+    if (MO.isGlobal()) {
+      GVSet.insert(MO.getGlobal());
+    } else if (MO.isReg() && MO.getReg().isVirtual()) {
+      const Register Reg = MO.getReg();
+      if (const MachineInstr *RegDef = MRI.getVRegDef(Reg))
+        findGlobalValues(*RegDef, GVSet, VisitedInstrs, RecMax - 1, MRI);
+    }
+  }
+}
+
+bool fixLoadMemOpInfo(MachineFunction &MF, MachineBasicBlock &MBB,
+                      MachineRegisterInfo &MRI) {
+  bool Changed = false;
+
+  for (MachineInstr &MI : MBB) {
+    if (!MI.mayLoad() || !MI.memoperands_empty())
+      continue;
+
+    SmallPtrSet<const Value *, 4> GVSet;
+    SmallPtrSet<const MachineInstr *, 4> VisitedInstrs;
+
+    // As we don't know which operand is the pointer, we need
+    // to iterate over all uses.
+    for (auto MO : MI.uses()) {
+      if (MO.isReg() && MO.getReg().isVirtual()) {
+        const Register Reg = MO.getReg();
+        if (const MachineInstr *RegDef = MRI.getVRegDef(Reg))
+          findGlobalValues(*RegDef, GVSet, VisitedInstrs, 10, MRI);
+      }
+    }
+
+    // Add gattered GlobalValues as MMOs.
+    for (auto GV : GVSet) {
```

Comment:
I think we should know whether we completed the search, otherwise we may be adding an incomplete set, which can cause over-optimistic alias analysis